clippy_lints/src/no_mangle_with_rust_abi.rs

```diff
@@ -37,33 +45,79 @@ declare_lint_pass!(NoMangleWithRustAbi => [NO_MANGLE_WITH_RUST_ABI]);
+            ItemKind::Static(ty, _, _) => check_static(cx, item, ty),
+            _ => {},
+        }
+    }
+}
 
-                    span_lint_and_then(
-                        cx,
-                        NO_MANGLE_WITH_RUST_ABI,
-                        fn_sig.span,
-                        "`#[no_mangle]` set on a function with the default (`Rust`) ABI",
-                        |diag| {
-                            diag.span_suggestion(sugg_span, "set an ABI", "extern \"C\" ", app)
-                                .span_suggestion(sugg_span, "or explicitly set the default", "extern \"Rust\" ", app);
-                        },
+/// Check for functions that are implicitly using the Rust ABI.
+fn check_fn(cx: &LateContext<'_>, fn_sig: FnSig<'_>) {
+    if fn_sig.header.abi == Abi::Rust
+        && let Some(snippet) = snippet_opt(cx, fn_sig.span)
+        && let Some((fn_attrs, _)) = snippet.split_once("fn")
+        && !fn_attrs.contains("extern")
+    {
+        let sugg_span = fn_sig
+            .span
+            .with_lo(fn_sig.span.lo() + BytePos::from_usize(fn_attrs.len()))
+            .shrink_to_lo();
+
+        span_lint_and_then(
+            cx,
+            NO_MANGLE_WITH_RUST_ABI,
+            fn_sig.span,
+            "`#[no_mangle]` set on a function with the default (`Rust`) ABI",
+            |diag| {
+                diag.note_once(
+                    "calling this function in an external content is Undefined Behavior, as its calling convention \
+                     (ABI) is undefined",
+                )
+                .span_suggestion(sugg_span, "set an ABI", "extern \"C\" ", Applicability::MaybeIncorrect)
+                .span_suggestion(
+                    sugg_span,
+                    "or explicitly set the default",
+                    "extern \"Rust\" ",
+                    Applicability::MaybeIncorrect,
+                );
+            },
+        );
+    }
+}
+
+/// Check for static items with a type that is implicitly using the Rust ABI.
+fn check_static(cx: &LateContext<'_>, item: &Item<'_>, ty: &Ty<'_>) {
+    // TODO(Centri3): Add a check here for `#[repr(Rust)]` once #114201 is merged.
+
+    if let Some(is_local) = match cx.tcx.type_of(item.owner_id).skip_binder().kind() {
+        ty::Adt(def, _) => {
+            let repr = def.repr();
+            (!repr.c() && !repr.transparent() && !repr.simd() && repr.int.is_none()).then(|| def.did().is_local())
+        },
+        ty::Tuple(tys) => (tys.len() != 0).then_some(false),
+        _ => return,
+    } {
+        span_lint_and_then(
+            cx,
+            NO_MANGLE_WITH_RUST_ABI,
+            item.span,
+            "`#[no_mangle]` set on a static with the default (`Rust`) ABI",
+            |diag| {
+                diag.span_note(ty.span, "this type is implicitly `#[repr(Rust)]`")
+                    .note_once(
+                        "usage of this static in an external context is Undefined Behavior, as its layout is undefined",
```

Comment:
Don't think this note is necessary, but I think the additional context could be nice.

Reply:
The note is technically wrong though (undefined layout instead of unstable layout), that should be changed

Reply:
Is UB the right term for this? It might lead to UB if it causes a read from padding bytes for example but I don't imagine any usage of it at all is UB

e.g. it might be a valid case to pass a pointer to a rust type to other functions exposed by a rust crate, there's also the case of rust to rust FFI when building both sides with the exact same toolchain

Reply:
> It might lead to UB if it causes a read from padding bytes for example

Also if it rearranges the fields, lets say one field has an invariant that it's always even, but the other doesn't, and the compiler's aware of this invariant; if you pass that to a function expecting them to *not* be rearranged, what will happen?

I think it's the right term, any use of it inside of Rust isn't defined behavior, its layout can be whatever it wants, outside of rust it isn't really UB, but I think it still conveys the point better than explicitly stating this.

> when building both sides with the exact same toolchain

Doesn't mean their layouts are the same, it's still unstable even with the same compiler flags and all.

Reply:
> it's still unstable even with the same compiler flags and all

It is a grey area but it's still done, rustc and clippy dynamically link to libstd/librustc_driver for example

> outside of rust it isn't really UB, but I think it still conveys the point better than explicitly stating this

UB is a term that is so often misunderstood and I don't think we should worsen that. It's fine to just say that the layout is unspecified

Reply:
> It is a grey area but it's still done, rustc and clippy dynamically link to libstd/librustc_driver for example

Those are `dylib`s. They have the necessary info to ensure everything works fine, though whether or not you can actually read that info is up to whether they're unstable even between the same compiler version (afaik, they are, but in practice they may work fine between the same compiler version. Otherwise you wouldn't be able to link to libstd anyway)

This mostly concerns `cdylib`. If you're using `dylib`, there's no point to using `no_mangle` because you can just access the static idiomatically using a single path.

Reply:
I'm aware it's not the same situation, it's just a grey area example